components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -284,51 +293,91 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
-                    // In the first iteration resultedUserList will be empty.
-                    if (ArrayUtils.isNotEmpty(resultedUserList)) {
-                        // Remove the users from already matched list who are not in the recently matched list.
-                        resultedUserList = getCommonUserEntries(resultedUserList, matchedUserList, key, value);
-                        if (ArrayUtils.isEmpty(resultedUserList)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("No user matched for given claims");
-                            }
-                            return StringUtils.EMPTY;
-                        }
-                    } else {
-                        resultedUserList = matchedUserList;
-                    }
-                } else {
+                Condition operationalCondition = getOperationalCondition(expressionConditionList);
+                /* Get the users list that matches with the condition
+                   limit : 2, offset : 1, sortBy : null, sortOrder : null */
+                resultedUserList.addAll(abstractUserStoreManager.getUserListWithID(operationalCondition, domain,
+                        UserCoreConstants.DEFAULT_PROFILE, 2, 1, null, null));
+                if (resultedUserList.size() > 1) {
                     if (log.isDebugEnabled()) {
-                        String message = String.format("No users matched for claim : %s for value : %s", key, value);
-                        log.debug(message);
+                        log.debug("Multiple users matched for given claims set : " +
+                                Arrays.toString(resultedUserList.toArray()));
                     }
-                    return StringUtils.EMPTY;
+                    throw Utils.handleClientException(
+                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_MULTIPLE_MATCHING_USERS, null);
                 }
             }
-        }
-        // Return matched user.
-        if (ArrayUtils.isNotEmpty(resultedUserList) && resultedUserList.length == 1) {
-            return resultedUserList[0];
-        } else {
+            // Return empty when no users are found.
+            if (resultedUserList.isEmpty()) {
+                return StringUtils.EMPTY;
+            }
+            // When the code reaches here there only be single user match.
+            return resultedUserList.get(0).getDomainQualifiedUsername();
+        } catch (org.wso2.carbon.user.core.UserStoreException e) {
             if (log.isDebugEnabled()) {
-                log.debug("Multiple users matched for given claims set : " + Arrays.toString(resultedUserList));
+                log.debug("Error while getting users from user store for the given claim set: " +
+                        Arrays.toString(claims.keySet().toArray()));
+            }
+            throw new IdentityRecoveryException(e.getErrorCode(), e.getMessage(), e);
```

Comment:
Check whether we can define a specific error message than passing the message coming in the exception

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/631fed769ea97d4dd15070b776d4126eab1a80d6